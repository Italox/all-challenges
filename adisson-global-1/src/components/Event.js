import React, { useState, useEffect } from 'react';
import { Pane, Card, Heading, Button } from 'evergreen-ui';

export default function Event({ id, name, markets }) {
  const [checked, setChecked] = useState([]);

  useEffect(() => {
    const bets = localStorage.getItem('bets', JSON.stringify(checked));
    if (bets) {
      setChecked(bets);
    }
  }, [checked]);

  function changeButton(eventId, marketName, betName, betValue) {
    const betsStorage = JSON.parse(localStorage.getItem('bets'));

    if (betsStorage) {
      if (betsStorage[eventId]) {
        if (betsStorage[eventId][marketName]) {
          betsStorage[eventId][marketName] = {
            [betName]: betValue,
          };
        } else {
          betsStorage[eventId] = {
            ...betsStorage[eventId],
            [marketName]: { [betName]: { betValue } },
          };
        }
        const updateStorage = {
          ...betsStorage,
        };

        localStorage.setItem('bets', JSON.stringify(updateStorage));
        setChecked(JSON.stringify(updateStorage));
        return;
      }

      const updateStorage = {
        ...betsStorage,
        [eventId]: { [marketName]: { [betName]: { betValue } } },
      };

      localStorage.setItem('bets', JSON.stringify(updateStorage));
      setChecked(JSON.stringify(updateStorage));
      return;
    }

    localStorage.setItem(
      'bets',
      JSON.stringify({
        [eventId]: { [marketName]: { [betName]: { betValue } } },
      })
    );
  }

  return (
    <Card key={id} border="default" marginTop={15} paddingBottom={20}>
      <Pane
        background="tint1"
        paddingTop={15}
        paddingBottom={15}
        display="flex"
        alignItems="center"
        justifyContent="center"
      >
        <Heading is="h1">{name}</Heading>
      </Pane>
      {markets.map(market => (
        <Pane key={market.id} paddingBottom={10} paddingTop={20}>
          <Heading is="h4" paddingBottom={20} paddingLeft={20}>
            {market.name}
          </Heading>
          <Pane
            width="100%"
            display="flex"
            alignItems="left"
            paddingLeft={20}
            justifyContent="left"
            flexWrap="wrap"
          >
            {market.selections.map(option => (
              <Button
                key={option.id}
                height={32}
                marginRight={10}
                marginTop={5}
                intent="success"
                appearance={
                  checked.includes(option.name) ? 'primary' : 'default'
                }
                onClick={() =>
                  changeButton(id, market.name, option.name, option.price)
                }
              >
                {option.name} - {option.price}
              </Button>
            ))}
          </Pane>
        </Pane>
      ))}
    </Card>
  );
}
