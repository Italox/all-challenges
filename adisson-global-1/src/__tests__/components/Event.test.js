import React from 'react';
import { render, fireEvent } from '@testing-library/react';
import Event from '../../components/Event';

const marketsModel = [
  {
    id: 'MKT_1',
    name: 'Team to Win',
    selections: [
      {
        id: 'SEL_1',
        name: 'Real Madrid',
        price: 1.23,
      },
      {
        id: 'SEL_2',
        name: 'Barcelona',
        price: 2.05,
      },
    ],
  },
];

const addBetMock = jest.fn();
const checkedBets = [
  {
    marketName: 'Team to Win',
    betId: 'SEL_1',
    betName: 'Real Madrid',
    betPrice: 1.23,
  },
];

describe('Test Event Component', () => {
  beforeEach(() => {
    localStorage.clear();
    jest.clearAllMocks();
  });

  it('should render event', () => {
    const { getByText } = render(
      <Event
        key="EVT_1"
        id="EVT_1"
        name="Real Madrid vs Barcelona"
        markets={marketsModel}
        checkedBets={[]}
      />
    );

    expect(getByText('Team to Win')).toBeInTheDocument();
  });

  it('change button color when click', () => {
    const { getByText, rerender } = render(
      <Event
        id="EVT_1"
        name="Real Madrid vs Barcelona"
        markets={marketsModel}
        addBet={addBetMock}
        checkedBets={[]}
      />
    );
    const marketButton = getByText('Real Madrid - 1.23');
    fireEvent.click(marketButton);
    expect(addBetMock).toBeCalledTimes(1);
    rerender(
      <Event
        id="EVT_1"
        name="Real Madrid vs Barcelona"
        markets={marketsModel}
        checkedBets={checkedBets}
      />
    );
    expect(marketButton).toHaveStyle(`
      background-image: linear-gradient(rgb(35, 194, 119), rgb(57, 157, 108));
      color:white;
    `);
  });
});
