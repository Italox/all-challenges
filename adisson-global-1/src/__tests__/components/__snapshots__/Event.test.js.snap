// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render correct event 1`] = `
<DocumentFragment>
  <div
    class="css-nil ðŸ“¦b-top_1px-solid-E4E7EB ðŸ“¦b-rgt_1px-solid-E4E7EB ðŸ“¦b-btm_1px-solid-E4E7EB ðŸ“¦b-lft_1px-solid-E4E7EB ðŸ“¦bblr_5px ðŸ“¦bbrr_5px ðŸ“¦btlr_5px ðŸ“¦btrr_5px ðŸ“¦mt_15px ðŸ“¦pb_20px ðŸ“¦box-szg_border-box"
  >
    <div
      class="css-nil ðŸ“¦bg_1vd85pq ðŸ“¦pt_15px ðŸ“¦pb_15px ðŸ“¦dspl_flex ðŸ“¦algn-itms_center ðŸ“¦just-cnt_center ðŸ“¦box-szg_border-box"
    >
      <h1
        class="ðŸ“¦mt_0px ðŸ“¦mb_0px ðŸ“¦fnt-sze_16px ðŸ“¦f-wght_500 ðŸ“¦ln-ht_20px ðŸ“¦ltr-spc_-0-05px ðŸ“¦fnt-fam_b77syt ðŸ“¦color_234361 ðŸ“¦box-szg_border-box"
      >
        Real Madrid vs Barcelona
      </h1>
    </div>
    <div
      class="css-nil ðŸ“¦pb_10px ðŸ“¦pt_20px ðŸ“¦box-szg_border-box"
    >
      <h4
        class="ðŸ“¦mt_0px ðŸ“¦mb_0px ðŸ“¦fnt-sze_16px ðŸ“¦f-wght_500 ðŸ“¦ln-ht_20px ðŸ“¦ltr-spc_-0-05px ðŸ“¦fnt-fam_b77syt ðŸ“¦color_234361 ðŸ“¦pb_20px ðŸ“¦pl_20px ðŸ“¦box-szg_border-box"
      >
        Team to Win
      </h4>
      <div
        class="css-nil ðŸ“¦w_100prcnt ðŸ“¦dspl_flex ðŸ“¦algn-itms_left ðŸ“¦pl_20px ðŸ“¦just-cnt_left ðŸ“¦flx-wrap_wrap ðŸ“¦box-szg_border-box"
      >
        <button
          class="css-11fu0id ðŸ“¦fnt-fam_b77syt ðŸ“¦mt_5px ðŸ“¦fnt-sze_12px ðŸ“¦f-wght_500 ðŸ“¦ln-ht_32px ðŸ“¦ltr-spc_0 ðŸ“¦btrr_3px ðŸ“¦bbrr_3px ðŸ“¦btlr_3px ðŸ“¦bblr_3px ðŸ“¦pt_0px ðŸ“¦pb_0px ðŸ“¦pr_16px ðŸ“¦pl_16px ðŸ“¦ml_0px ðŸ“¦mr_10px ðŸ“¦mb_0px ðŸ“¦h_32px ðŸ“¦pst_relative ðŸ“¦dspl_inline-flex ðŸ“¦algn-itms_center ðŸ“¦flx-wrap_nowrap ðŸ“¦box-szg_border-box"
        >
          Real Madrid - 1.23
        </button>
        <button
          class="css-11fu0id ðŸ“¦fnt-fam_b77syt ðŸ“¦mt_5px ðŸ“¦fnt-sze_12px ðŸ“¦f-wght_500 ðŸ“¦ln-ht_32px ðŸ“¦ltr-spc_0 ðŸ“¦btrr_3px ðŸ“¦bbrr_3px ðŸ“¦btlr_3px ðŸ“¦bblr_3px ðŸ“¦pt_0px ðŸ“¦pb_0px ðŸ“¦pr_16px ðŸ“¦pl_16px ðŸ“¦ml_0px ðŸ“¦mr_10px ðŸ“¦mb_0px ðŸ“¦h_32px ðŸ“¦pst_relative ðŸ“¦dspl_inline-flex ðŸ“¦algn-itms_center ðŸ“¦flx-wrap_nowrap ðŸ“¦box-szg_border-box"
        >
          Barcelona - 2.05
        </button>
      </div>
    </div>
    <div
      class="css-nil ðŸ“¦pb_10px ðŸ“¦pt_20px ðŸ“¦box-szg_border-box"
    >
      <h4
        class="ðŸ“¦mt_0px ðŸ“¦mb_0px ðŸ“¦fnt-sze_16px ðŸ“¦f-wght_500 ðŸ“¦ln-ht_20px ðŸ“¦ltr-spc_-0-05px ðŸ“¦fnt-fam_b77syt ðŸ“¦color_234361 ðŸ“¦pb_20px ðŸ“¦pl_20px ðŸ“¦box-szg_border-box"
      >
        Player to Score First
      </h4>
      <div
        class="css-nil ðŸ“¦w_100prcnt ðŸ“¦dspl_flex ðŸ“¦algn-itms_left ðŸ“¦pl_20px ðŸ“¦just-cnt_left ðŸ“¦flx-wrap_wrap ðŸ“¦box-szg_border-box"
      >
        <button
          class="css-11fu0id ðŸ“¦fnt-fam_b77syt ðŸ“¦mt_5px ðŸ“¦fnt-sze_12px ðŸ“¦f-wght_500 ðŸ“¦ln-ht_32px ðŸ“¦ltr-spc_0 ðŸ“¦btrr_3px ðŸ“¦bbrr_3px ðŸ“¦btlr_3px ðŸ“¦bblr_3px ðŸ“¦pt_0px ðŸ“¦pb_0px ðŸ“¦pr_16px ðŸ“¦pl_16px ðŸ“¦ml_0px ðŸ“¦mr_10px ðŸ“¦mb_0px ðŸ“¦h_32px ðŸ“¦pst_relative ðŸ“¦dspl_inline-flex ðŸ“¦algn-itms_center ðŸ“¦flx-wrap_nowrap ðŸ“¦box-szg_border-box"
        >
          Ronaldo - 1.15
        </button>
        <button
          class="css-11fu0id ðŸ“¦fnt-fam_b77syt ðŸ“¦mt_5px ðŸ“¦fnt-sze_12px ðŸ“¦f-wght_500 ðŸ“¦ln-ht_32px ðŸ“¦ltr-spc_0 ðŸ“¦btrr_3px ðŸ“¦bbrr_3px ðŸ“¦btlr_3px ðŸ“¦bblr_3px ðŸ“¦pt_0px ðŸ“¦pb_0px ðŸ“¦pr_16px ðŸ“¦pl_16px ðŸ“¦ml_0px ðŸ“¦mr_10px ðŸ“¦mb_0px ðŸ“¦h_32px ðŸ“¦pst_relative ðŸ“¦dspl_inline-flex ðŸ“¦algn-itms_center ðŸ“¦flx-wrap_nowrap ðŸ“¦box-szg_border-box"
        >
          Messi - 1.3
        </button>
        <button
          class="css-11fu0id ðŸ“¦fnt-fam_b77syt ðŸ“¦mt_5px ðŸ“¦fnt-sze_12px ðŸ“¦f-wght_500 ðŸ“¦ln-ht_32px ðŸ“¦ltr-spc_0 ðŸ“¦btrr_3px ðŸ“¦bbrr_3px ðŸ“¦btlr_3px ðŸ“¦bblr_3px ðŸ“¦pt_0px ðŸ“¦pb_0px ðŸ“¦pr_16px ðŸ“¦pl_16px ðŸ“¦ml_0px ðŸ“¦mr_10px ðŸ“¦mb_0px ðŸ“¦h_32px ðŸ“¦pst_relative ðŸ“¦dspl_inline-flex ðŸ“¦algn-itms_center ðŸ“¦flx-wrap_nowrap ðŸ“¦box-szg_border-box"
        >
          Bale - 1.35
        </button>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
